src/js/BookReader.js

```diff
@@ -2721,6 +2725,50 @@ BookReader.prototype.canSwitchToMode = function(mode) {
   return true;
 };
 
+
+/**
+ * Returns the page URI or transparent image if out of range
+ * Also makes the reduce argument optional
+ * @param {number} index
+ * @param {number} [reduce]
+ * @param {number} [rotate]
+ * @return {string}
+ */
+BookReader.prototype._getPageURISrcset = function(index, reduce, rotate) {
+  let srcsetStr = "";
+  if (index < 0 || index >= this._models.book.getNumLeafs()) { // Synthesize page
+    return srcsetStr;
+  }
+
+  let ratio = reduce;
+  if ('undefined' == typeof(reduce)) {
+    // reduce not passed in
+    // $$$ this probably won't work for thumbnail mode
+    ratio = this._models.book.getPageHeight(index) / this.twoPage.height;
+  }
+  var scale;
+  // $$$ we make an assumption here that the scales are available pow2 (like kakadu)
+  if (ratio < 2) {
+    return srcsetStr;
+  } else if (ratio < 4) {
+    scale = [1];
+  } else if (ratio < 8) {
+    scale = [2,1];
+  } else if (ratio < 16) {
+    scale = [4,2,1];
+  } else  if (ratio < 32) {
+    scale = [8,4,2,1];
+  } else {
+    scale = [16,8,4,2,1];
+  }
+  scale.forEach((el, i) => {
```

Comment:
Rather than performing string concatenation with each iteration of this loop, you can instead use `scale.map` to construct an array of sources and then `.join` them. This will also allow you to avoid having to append the final ", " to the string. So something like:

```js
return scale.map((el, i) => (
  this._models.book.getPageURI(index, scale[i], rotate) + " "
      + Math.pow(2, i + 1) + "x"
)).join(', ');
```

Reply:
done